stlab/future.hpp

```diff
@@ -814,23 +814,24 @@ struct when_all_shared {
     std::tuple<boost::optional<Ts>...>  _args;
     future<void>                        _holds[sizeof...(Ts)] {};
     std::atomic_size_t                  _remaining {sizeof...(Ts)};
-    std::atomic_bool                    _error_happend{ false };
-    std::mutex                          _errormutex;
+    std::atomic_bool                    _error_happened{ false };
     boost::optional<std::exception_ptr> _error;
     packaged_task<>                     _f;
 
-    void done() { if (--_remaining == 0 && !_error_happend) _f(); }
+    void done() { if (!_error_happened && --_remaining == 0) _f(); }
 
     void failure(std::exception_ptr error) {
-        {
-            std::unique_lock<std::mutex> lock(_errormutex);
+        bool current_error_happened = _error_happened.load();
```

Comment:
I am not sure, if this code snippet is just for illustration purpose. (But passing false as first parameter does not compile, because the 1st parameter must be passed by value.)
I tried to remove the check above, but then the software crashes within 1 : 50 runs. When I leave it in the code I have no crash within 5000 runs. Hmmm.
I stepped through the assembly and as far as I see it, compare_exchange_strong returns always true when the value is valid to be set. A look into the Intel docs of cmpxchg suggest something similar:

TEMP <- DEST
IF accumulator = TEMP
  THEN
    ZF <- 1;
    DEST <- SRC;
  ELSE
    ZF <- 0
    accumulator <- TEMP
    DEST <- TEMP

So ZF, which is used by the following jump, is always set.

So this results then in a double call of _f() which is nooot good.
So I leave the checks in the code.